### ISSUE TITLE
Using ALPSO in parallel mode (ALPSO: IndexError: too many indices for array)

### ISSUE THREAD

**Issue by kmodexc** [mdolab/pyoptsparse#390]
## Description
Error when executing simple optimization using ALPSO.

### Steps to reproduce issue

```
from pyoptsparse import OPT, Optimization
import numpy as np

class PyOptSparseModel():
    def __init__(self):
        self.optProb = Optimization("DS Opt", self.objfunc)
        self.optProb.addVarGroup("var", 2, "c")
        self.optProb.addObj("y")

    def objfunc(self,xdict):
        x = xdict["var"]

        funcs = {}
        funcs["y"] = np.sum(x)

        fail = False
        return funcs, fail
    
    def optimize(self, train_steps=10):

        options = {}
        options["ALPSO"] = { "maxOuterIter": train_steps, "printOuterIters": 1, "parallelType": "EXT", "filename": "/work/ALPSO.out" }

        opt = OPT("ALPSO",options=options["ALPSO"])
        sol = opt(self.optProb)

        self.solution = sol.getDVs()

        return self.solution

if __name__ == "__main__":
    model = PyOptSparseModel()
    solution = model.optimize(train_steps=10)
    print(solution)
```

### Current behavior
```
Traceback (most recent call last):
  File "/work/alpso_test.py", line 33, in <module>
    solution = model.optimize(train_steps=10)
  File "/work/alpso_test.py", line 25, in optimize
    sol = opt(self.optProb)
  File "/opt/pyoptsparse/pyoptsparse/pyALPSO/pyALPSO.py", line 172, in __call__
    opt_x, opt_f, opt_g, opt_lambda, nfevals, rseed = self.alpso.alpso(
  File "/opt/pyoptsparse/pyoptsparse/pyALPSO/alpso_ext.py", line 232, in alpso
    L[i] = f[i]
IndexError: too many indices for array: array is 0-dimensional, but 1 were indexed

```

### Code versions
- Operating System: Ubuntu 22.04.4
- Python: 3.10.12
- This repository: [cae6954](https://github.com/mdolab/pyoptsparse/commit/cae6954e67ad58b03b535cc015e9d0118d89695e)

**Comment by marcomangano** [comment]
Hello @kmodexc, thanks for sharing a MWE, I could replicate your error.
I initially thought this was a bug, but if you `diff` `alpso.py` and `alpso_ext` (the latter is the one imported when `"parallelType": "EXT"`) you will see how the parallel version expects a different `objFunc` signature that returns a vector `f` of `len(swarsize)` instead of a scalar.
I am not familiar with this PSO algorithm , but in the default mode [`alpso` loops over every particle of the swarm](https://github.com/mdolab/pyoptsparse/blob/cae6954e67ad58b03b535cc015e9d0118d89695e/pyoptsparse/pyALPSO/alpso.py#L217-L229) and evaluates the (scalar) objfunc at that point. In the parallel version, [there is no such loop](https://github.com/mdolab/pyoptsparse/blob/cae6954e67ad58b03b535cc015e9d0118d89695e/pyoptsparse/pyALPSO/alpso_ext.py#L217-L227). We do not have examples afaik, but it looks like `alpso_ext` passes an array with all the DVs of all the particles at a given iteration, and expects a vector f with all the function values of the particles at the same iteration.

In short, if you remove the `"parallelType": "EXT"` option the optimizer will work. To run the optimizer in parallel mode, it is on the user to provide an `objfunc` that runs in parallel and returns a vector of the correct size.
